### PR TITLE
Move template registry from AdminBundle

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

Template registry is templates manager  fuature and will be nice to allow use it without `AdminBundle`. This PR will copy `Templating` scructure to `twig-extensions` repository.

I am targeting this branch, because this change respect BC.

## Changelog

```markdown
### Added
- Added `Sonata\AdminBundle\Templating\AbstractTemplateRegistry`
- Added `Sonata\AdminBundle\Templating\MutableTemplateRegistry`
- Added `Sonata\AdminBundle\Templating\MutableTemplateRegistryAwareInterface`
- Added `Sonata\AdminBundle\Templating\MutableTemplateRegistryInterface`
- Added `Sonata\AdminBundle\Templating\TemplateRegistry`
- Added `Sonata\AdminBundle\Templating\TemplateRegistryAwareInterface`
- Added `Sonata\AdminBundle\Templating\TemplateRegistryInterface`
```

## To do
- [ ] Update the documentation;

### CODE REVIEW

**Comment by core23** [comment]
Thinking a little more about this. What's the actual purpose of this general registry?

If you want to override a template, you should use the existing twig mechanism. IMHO we only need this for templates inside the admin context.

**Comment by wbloszyk** [comment]
I do not wanna create general registry. 

**My idea is:**
- move Templating classes to twig-extensions (people can create registry by own - like AdminBundle)

**And as seperate feature:**
- add template registry pool (also as twig extensions with global var)
- add `support()` method to allow use multiple registry for one context

**Then in AdminBundle:**  
- we can remove current annoying `Admin` - `TemplateRegistry` relation and inject `TemplateRegistryPool` insead
- pool will autogenerate (copy) registry for admin
- registry service for `Admin` will be optional, it can be get from pool

**I know we are using it only with admin context but IMHO:**
- we should keep code by functionality
- template registry can be stanalone feature (we dont need Admin code in it)
- i need this feature in m next project :smile: 
- will be nice make major release for twig-extensions (toremove hack from flashManager and improve template registry) 

@core23 
I hope I explain very well why I want move this functionallity.

**Comment by VincentLanglet** [comment]
> - i need this feature in m next project :smile: 

Can you give us the usage/context ? It could help to understand the purpose of this feature in another use case than the AdminBundle

**Comment by wbloszyk** [comment]
> > * i need this feature in m next project smile
> 
> Can you give us the usage/context ? It could help to understand the purpose of this feature in another use case than the AdminBundle

**What template registry is?**
TemplateRegistry is configurable object using instead static template name. It mean template can be dynamicly change.

**TemplateRegistry in AdminBundle.**
In AdminBundle we are define templates by `name` => `value` where `name` is equal to type. 
For example: `'show' => '@FooAdmin/CRUD/show.hml.twig'`

TemplateRegistry in other project.
- **Case 1**
     shop - category catalog - diffrence list for some category
    `'show' => '@FooShop/Catalog/show.hml.twig'` 
- **Case 2**
    shop - product view on list - multiple view 
```php
    'show' => [ 
             'default' => '@FooShop/product/thumbnail.hml.twig',
             'thumbnail' => '@FooShop/product/thumbnail.hml.twig',
             'thumbnail_table' => '@FooShop/product/thumbnail_table.hml.twig',
             'thumbnail_list' => '@FooShop/product/thumbnail_list.hml.twig',  
    ] 
```
- **Case 3**
    building projects - a lot of building types like house, markets, dog house ... can have many optional things like garage, terrace ... Pool will allow get registry for example GarageAware (in twig), and display garage in tab.

- **Other cases**
    CMS, using diffrenct registry to use diffrence css frameworks, ...

**Review by VincentLanglet — APPROVED** [review]

File: src/Templating/MutableTemplateRegistryAwareInterface.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @author Wojciech Błoszyk <wbloszyk@gmail.com>
+ *
+ * @method MutableTemplateRegistryInterface getTemplateRegistry()
+ * @method bool                             hasTemplateRegistry()
+ * @method void                             setTemplateRegistry(MutableTemplateRegistryInterface $templateRegistry)
+ */
+interface MutableTemplateRegistryAwareInterface
+{
+    // NEXT_MAJOR: uncomment this method
```
Comment (core23) [inline src/Templating/MutableTemplateRegistryAwareInterface.php:25]: It's a new interface, you should add this methods now. Do not mind the admin bundle when adding a new interface.

File: src/Templating/MutableTemplateRegistryInterface.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+     * NEXT_MAJOR: remove this method declaration with docblock and uncomment code below.
+     *
+     * @param array<string, string> $templates 'name' => 'file_path.html.twig'
+     */
+    public function setTemplates(array $templates);
+
+    ///**
+    // * @param array<string, string> $templates 'name' => 'file_path.html.twig'
+    // */
+    //public function setTemplates(array $templates): void;
```
Comment (core23) [inline src/Templating/MutableTemplateRegistryInterface.php:31]: Same here

File: src/Templating/TemplateRegistry.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\Twig\Templating;
+
+/**
+ * @author Timo Bakx <timobakx@gmail.com>
+ *
+ * NEXT_MAJOR: remove `MutableTemplateRegistryInterface` implementation.
```
Comment (core23) [inline src/Templating/TemplateRegistry.php:19]: And here

File: src/Templating/TemplateRegistryAwareInterface.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @author Wojciech Błoszyk <wbloszyk@gmail.com>
+ *
+ * @method TemplateRegistryInterface getTemplateRegistry()
+ * @method bool                      hasTemplateRegistry()
+ * @method void                      setTemplateRegistry(TemplateRegistryInterface $templateRegistry)
+ */
+interface TemplateRegistryAwareInterface
+{
+    // NEXT_MAJOR: uncomment this method
```
Comment (core23) [inline src/Templating/TemplateRegistryAwareInterface.php:25]: And here... and on the other classes.

**Review by core23 — CHANGES_REQUESTED** [review]
Please add an example doc so user can understand the purpose of the new template registry.

**Comment by core23** [comment]
> I hope I explain very well why I want move this functionallity.

Given the use cases where someone can use this, I can understand your idea. 

I'm still not sure if it's a good design idea to have a reusable template registry. There may be some coincidence with you actual use case now, but I have the fear that it might differ in the future.

I'm not completely against this feature, but I have come concerns.

File: src/Templating/MutableTemplateRegistryInterface.php (hunk 2 above)
Comment (wbloszyk) [reply src/Templating/MutableTemplateRegistryInterface.php:31]: I can't becouse it have to be the same as in AdminBundle:

https://github.com/sonata-project/SonataAdminBundle/pull/6755

Comment (VincentLanglet) [reply src/Templating/MutableTemplateRegistryInterface.php:31]: The solution would be to deprecate the ones in AdminBundle in the 4.x version then

**Pull request closed without merging**